sympy/stats/rv_interface.py

```diff
@@ -339,6 +340,44 @@ def factorial_moment(X, n, condition=None, **kwargs):
     """
     return expectation(FallingFactorial(X, n), condition=condition, **kwargs)
 
+def median(X, evaluate=True, **kwargs):
+    """
+    Calculuates the median of the probability distribution.
+
+    Parameters
+    ==========
+
+    X: The random expression whose median is to be calculated.
+
+    Returns
+    =======
+
+    The median of the random expression.
+
+    Examples
+    ========
+
+    >>> from sympy.stats import Normal, Die, median
+    >>> N = Normal('N', 3, 1)
+    >>> median(N)
+    3
+    >>> D = Die('D')
+    >>> median(D)
+    3
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Median#Probability_distributions
+
+    """
+    result = quantile(X, evaluate=evaluate, **kwargs)(Rational(1, 2))
+    if isinstance(result, FiniteSet):
```

Comment:
Unless I'm missing something, this doesn't make sense. A quantile is a number, not a set.

Reply:
```
>>> quantile(Beta('x',1,2))(Rational(1,2))
FiniteSet(1 - sqrt(2)/2)
```
I am getting the above output as FiniteSet, so I used the if block

Reply:
I think the Lambda function is returning the finite set after substituting the values

Reply:
I'm not sure what conventions are in cases that the median is not unique. e.g. the distribution:
```julia
In [8]: pdf = Piecewise((S.Half, (0<=x)&(x<1)), (0, (x>=1)&(x<2)), (S.Half, (x>=2)&(x<3)), (0, True))                                          

In [9]: pdf                                                                                                                                    
Out[9]: 
⎧1/2  for x ≥ 0 ∧ x < 1
⎪                      
⎪ 0   for x ≥ 1 ∧ x < 2
⎨                      
⎪1/2  for x ≥ 2 ∧ x < 3
⎪                      
⎩ 0       otherwise 
```
Depending on the definition of median any value in `Interval(1, 2)` can be a median.

Reply:
> 
> 
> I think the quantile is returning the finite set after substituting the values, while it should return an element.

Yes, `quantile` is a set. And `compute_quantile` is still returning it inside the Piecewise.  (See my previous comment concerning the `return` statement on line 305.) The `FiniteSet` with two elements that you originally pointed out only came about due to the incorrect domain parameter in the call to `solveset`.

As for @oscarbenjamin 's comment I think that a note should be added in `.median`'s docstring that clarifies the definition that is implicitly used (i.e. quantile = 0,5). It's good that you've chosen the `Die` example in the docstring as it helps illustrate this point.

Reply:
I will surely add that. But  for the quantile here, can also be a condition set, Also I found on the Wikipedia as the example produce by @oscarbenjamin :
"Any probability distribution on R has at least one median, but in specific cases there may be more than one median. Specifically, if a probability density is zero on an interval [a, b], and the cumulative distribution function at a is 1/2, any value between a and b will also be a median".
So raising error might not be a good option in that case instead I think we should return an element if the set is Singleton or esle the Finite Set. Am I correct?

Reply:
I too think that returning set would be a nice option though a Singleton or may be FiniteSet with more than one element. Is it a good idea to go for?

Reply:
Indexing is not supported in set.
For continuous interval I think it would return an interval.

Reply:
Are we sure there's not a misunderstanding here?

The set of possible median values has *nothing* to do with the sets currently returned by the `quantile` method. As per its current definition (see the docstring), `quantile` is supposed to return a single value (potentially as a piecewise), everything else is a bug.

*If* this new `median` method should return all possible values as a set (which would be fine by me), then the proposed implementation must be changed (see your own `Die` example in the docstring). Otherwise it'd be misleading to have `median` return an incomplete set. So IMO `median` should either return a single median value **or** the set of *all* median values.

Reply:
I checked again by just returning` quantile(X)(Rational(1,2))`, and I got m`edian(Die('D'))` as `3`, because Its domain is discrete.
Checking on to wikipedia I got to know that there can be an interval for median, as an example given by @oscarbenjamin , so I think we should return an` FiniteSet`, or an `Interval` if encountered,
As it is not affecting the` median` of `DiscreteDistribution` as I still get Median of Die as 3.
Should I go for `FiniteSet`?

Reply:
IMO, yes `FiniteSet` or ` Interval` are good.

Reply:
I will change it soon.